False positive `eslint@typescript-eslint/no-unused-vars` if local type is used in template declaration only

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I'm using [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue).
- [X] I'm sure the problem is a parser problem. (If you are not sure, search for the issue in [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue) repo and open the issue in [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue) repo if there is no solution.
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.

### What version of ESLint are you using?

8.39.0

### What version of `eslint-plugin-vue` and ` vue-eslint-parser` are you using?

- vue-eslint-parser@9.1.1
- eslint-plugin-vue@9.11.0


### What did you do?

<details>
<summary>Configuration</summary>

```
module.exports = {
  root: true,
  extends: [
    "plugin:vue/vue3-essential",
    "eslint:recommended",
    "@vue/eslint-config-typescript",
    "@vue/eslint-config-prettier",
  ],
  parser: "vue-eslint-parser",
  parserOptions: {
    parser: "@typescript-eslint/parser",
    ecmaVersion: "latest",
    sourceType: "module",
  },
};
```
</details>

```vue
<script setup lang="ts">
import HelloWorld from "./components/HelloWorld.vue";
import TheWelcome from "./components/TheWelcome.vue";
// 'Foo' is defined but never used. (eslint@typescript-eslint/no-unused-vars)
interface Foo {
  foo: string;
}
// 'Bar' is defined but never used. (eslint@typescript-eslint/no-unused-vars)
type Bar = string;
</script>

<template>
  <header>
    <img
      alt="Vue logo"
      class="logo"
      src="./assets/logo.svg"
      width="125"
      height="125"
    />

    <div class="wrapper">
      <HelloWorld msg="You did it!">
        <template #default="{ foo }: Foo">
          <p>Foo: {{ foo as Bar }}</p>
        </template>
      </HelloWorld>
    </div>
  </header>

  <main>
    <TheWelcome />
  </main>
</template>
```


### What did you expect to happen?

`Foo` and `bar` are used in template which should be a used var. No errors will be present.

### What actually happened?

```shell
> npm run lint      

> vue-eslint-unused-vars-repro@0.0.0 lint
> eslint . --ext .vue,.js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --fix --ignore-path .gitignore


C:\project\vue-eslint-unused-vars-repro\src\App.vue
  5:11  warning  'Foo' is defined but never used  @typescript-eslint/no-unused-vars
  9:6   warning  'Bar' is defined but never used  @typescript-eslint/no-unused-vars

✖ 2 problems (0 errors, 2 warnings)
```

### Link to Minimal Reproducible Example

https://github.com/Tanimodori/vue-eslint-unused-vars-repro/tree/local-define

**The repro is in the `local-define` branch.**

```shell
git clone https://github.com/Tanimodori/vue-eslint-unused-vars-repro
cd vue-eslint-unused-vars-repro
git checkout local-define
npm i
npm run lint
```

### Additional comments

This issue is similar to #174. The only difference is that this issue is occurred on local-defined types rather than imported types.

--- a/src/script-setup/scope-analyzer.ts
+++ b/src/script-setup/scope-analyzer.ts
@@ -205,6 +205,12 @@ function analyzeUsedInTemplateVariables(
 
         variable.references.push(reference)
         reference.resolved = variable
+
+        if (reference.isTypeReference) {
+            // @typescript-eslint/no-unused-vars treats type references at the same position as recursive references,
+            // so without this flag it will be marked as unused.
+            ;(variable as any).eslintUsed = true
+        }
     }
 
     function processVExpressionContainer(node: VExpressionContainer) {

--- a/test/fixtures/integrations/script-setup-with-typescript-eslint/no-unused-vars/.eslintrc.json
+++ b/test/fixtures/integrations/script-setup-with-typescript-eslint/no-unused-vars/.eslintrc.json
@@ -1,0 +1,5 @@
+{
+    "rules": {
+        "@typescript-eslint/no-unused-vars": "error"
+    }
+}

--- a/test/fixtures/integrations/script-setup-with-typescript-eslint/no-unused-vars/valid/type-def.vue
+++ b/test/fixtures/integrations/script-setup-with-typescript-eslint/no-unused-vars/valid/type-def.vue
@@ -1,0 +1,14 @@
+<script setup lang="ts">
+import MyComponent from "./components/MyComponent.vue";
+interface Foo {
+  foo?: string;
+}
+type Bar = string;
+</script>
+
+<template>
+  <MyComponent
+    :foo="{} as Foo"
+    :bar="'s' as Bar"
+  />
+</template>

